### PR TITLE
Ticket Price Calculator Enhancements

### LINKED ISSUE #2863: Ticket Price Calculator Enhancements
####  Add Default Taxes 

- in order to reduce the following:

     - time it takes users to configure prices for a ticket

     - errors caused by constantly re-entering price modifiers

     - duplicated records in the db

- we need to add a button just below the last row of the ticket price calculator labeled: :add default taxes"

- then load all of the default prices from the price table where `isDefault` and `isTax` are true ie: `PRC_is_default` === 1 && price type base type === 4

- set `isTaxable` to true on ticket

- we also do not want to clone default taxes and also NOT add relations in the ticket price table

- we also need to automagically load default taxes for tickets that already have `isTaxable` set to true

#### Free Tickets

- if a user doesn't require a price for a ticket, they should be able to skip the TPC step in the Edit/New Ticket Form Modal

- change existing button at bottom of Ticket Details step to read instead of "add ticket prices" 

- add a new button after it labeled "free ticket - assign dates" that skips the TPC and goes directly to TAM

#### But WAIT, There's MORE

- do NOT save prices to the db if their amount is 0

### PR DESCRIPTION
**Pull request by manzoorwanijk**
This PR refactors TPC to handle all the default price stuff client-side. Here is the list of changes in this PR:
- Adds support for price id param in prices GQL query
- Adds default prices to DOM data
- Enhances `usePrevNext` to add `goto` method
- Uses `FormSpy` in date/ticket edit form body to subscribe to form data for improved performance
- Improves ticket and price mutation in TPC
- Separates out logic by creating `useAddDefaultPrices`, `useAddDefaultTaxes`, `usePriceToTpcModifier` and the corresponding buttons.
- Creates `DeleteAllPricesButton` to nuke prices
- Creates No prices banner to show when there are no prices
- Improves TPC state management
- Adds price predicates related to `isDefault` prop

Closes #2863

### CODE REVIEW

File: assets/src/application/ui/display/confirm/useConfirmWithButton.tsx [hunk 1]
```diff
@@ -8,7 +8,7 @@ import { ConfirmPropsWithButton } from './types';
 import useConfirmationDialog from './useConfirmationDialog';
 
 const useConfirmWithButton: React.FC<ConfirmPropsWithButton> = ({ buttonProps, ...props }) => {
-	const confirmText = buttonProps?.tooltip ?? __('Please confirm this action.');
+	const confirmText = (props.title || buttonProps?.tooltip) ?? __('Please confirm this action.');
 	const { confirmationDialog, onOpen } = useConfirmationDialog({ ...props, confirmText });
 	const btnClassName = classNames(buttonProps.icon && iconBtnClassName, buttonProps.className);
```
Comment (manzoorwanijk) [inline assets/src/application/ui/display/confirm/useConfirmWithButton.tsx:13]: IMHO this needs some improvement

File: assets/src/domain/eventEditor/ui/datetimes/dateForm/multiStep/ContentBody.tsx [hunk 2]
```diff
@@ -15,40 +15,51 @@ const buttonRowStyle: CSSProperties = { display: 'flex', justifyContent: 'flex-e
...
+										onClick={form.submit}
+										isDisabled={isSubmitDisabled}
+									/>
+								</>
+							)}
+						</div>
+					</div>
+				);
+			}}
+		</FormSpy>
```
Comment (manzoorwanijk) [inline assets/src/domain/eventEditor/ui/datetimes/dateForm/multiStep/ContentBody.tsx:62]: `FormSpy` makes things awesome when you lockdown RFF re-rendering

File: assets/src/domain/eventEditor/ui/tickets/ticketForm/multiStep/ContentBody.tsx [hunk 3]
```diff
@@ -17,54 +17,78 @@ const buttonRowStyle: CSSProperties = { display: 'flex', justifyContent: 'flex-e
...
-	const prices = form.getState().values.prices || [];
-	const isTPCSubmitDisabled = prices.length && prices.some(({ amount }) => anyPass([isNil, isEmpty])(amount));
-
+	const subscription = { submitting: true, hasValidationErrors: true, hasSubmitErrors: true };
 	return (
-		<div>
-			<TicketFormSteps current={current} />
-			{/* RFF fields */}
-			{current === 0 && children}
+		<FormSpy subscription={subscription}>
```
Comment (manzoorwanijk) [inline assets/src/domain/eventEditor/ui/tickets/ticketForm/multiStep/ContentBody.tsx:29]: `FormSpy` simplified helped to simplify the things here.

File: assets/src/domain/eventEditor/ui/tickets/ticketForm/multiStep/ContentBody.tsx [hunk 4]
```diff
@@ -17,54 +17,78 @@ const buttonRowStyle: CSSProperties = { display: 'flex', justifyContent: 'flex-e
...
+										leftIcon='chevron-left'
+									/>
+									<Button buttonText={__('Previous')} onClick={prev} leftIcon='chevron-left' />
+									<Button
+										buttonText={__('Submit')}
+										onClick={form.submit}
+										isDisabled={isSubmitDisabled}
+									/>
+								</>
+							)}
```
Comment (manzoorwanijk) [inline assets/src/domain/eventEditor/ui/tickets/ticketForm/multiStep/ContentBody.tsx:86]: May be these buttons should be extracted in another PR.

File: assets/src/domain/eventEditor/ui/tickets/ticketPriceCalculator/data/useInitialState.ts [hunk 5]
```diff
@@ -39,28 +32,20 @@ const useInitialState = ({ ticketId }: BaseProps): StateInitializer => {
...
-	if (!hasBasePrice) {
-		// add the base price with `isNew` flag to make sure it's created on submit
-		// `order` as 1 to make sure it remains at the top
-		prices = [{ ...basePrice, order: 1, isNew: true }, ...prices];
-	}
-
 	return useCallback<StateInitializer>(
 		(initialState) => {
-			return { ...initialState, ticket: ticket ?? {}, prices };
+			return { ...initialState, ticket, prices };
```
Comment (manzoorwanijk) [inline assets/src/domain/eventEditor/ui/tickets/ticketPriceCalculator/data/useInitialState.ts:48]: State initialization is much simpler now

File: assets/src/domain/eventEditor/ui/tickets/ticketPriceCalculator/hooks/useMutatePrices.ts [hunk 6]
```diff
@@ -15,47 +15,51 @@ const useMutatePrices = (): Callback => {
...
-					if (!(isNew || isModified) && !price.isBasePrice) {
-						// retain the existing relation
-						relatedPriceIds.push(price.id);
-						return Promise.resolve(price);
-					}
-					const normalizedPriceFields = cloneAndNormalizePrice(price);
-					// if it's a newly added price
-					if (isNew) {
+			if (prices?.length) {
+				// make sure to complete all price mutatons before updating the ticket
```
Comment (manzoorwanijk) [inline assets/src/domain/eventEditor/ui/tickets/ticketPriceCalculator/hooks/useMutatePrices.ts:19]: This was needed for tickets without prices.

File: assets/src/domain/eventEditor/ui/tickets/ticketPriceCalculator/hooks/usePriceToTpcModifier.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+			// "priceType" and "priceTypeOrder"
+			return {
+				...price,
+				priceType: priceTypeId,
+				priceTypeOrder: priceTypeIdOrder[priceTypeId],
+			};
+		},
+		[getRelations, priceTypeIdOrder]
+	);
+};
```
Comment (manzoorwanijk) [inline assets/src/domain/eventEditor/ui/tickets/ticketPriceCalculator/hooks/usePriceToTpcModifier.ts:33]: I don't know why this wasn't created before 🤔

File: assets/src/domain/eventEditor/ui/tickets/ticketPriceCalculator/stateListeners/useInitStateListeners.ts [hunk 8]
```diff
@@ -17,8 +17,9 @@ const useStateListeners = (): void => {
 	}, [getData, setPrices]);
 
 	const updateTicketTotal = useCallback(() => {
-		const ticketTotal = calculateTicketTotal(getData());
-		updateTicketPrice(parsedAmount(formatAmount(ticketTotal)));
+		let ticketTotal = calculateTicketTotal(getData());
+		ticketTotal = parsedAmount(formatAmount(ticketTotal));
+		updateTicketPrice(isNaN(ticketTotal) ? 0 : ticketTotal);
```
Comment (manzoorwanijk) [inline assets/src/domain/eventEditor/ui/tickets/ticketPriceCalculator/stateListeners/useInitStateListeners.ts:22]: `parsedAmount` returns `NaN` for empty string

File: core/domain/services/admin/events/editor/EventEntityRelations.php [hunk 9]
```diff
@@ -123,8 +122,22 @@ private function processPrices(array $ticketIds)
...
+                    'Ticket.TKT_ID' => ['IN', $ticketIds],
+                    // or it's a default price and not trashed
+                    'AND' => [
+                        'PRC_deleted'    => 0,
+                        'PRC_is_default' => 1,
+                    ],
+                ],
+            ],
+            'group_by'                 => 'PRC_ID',
+            'default_where_conditions' => 'minimum',
```
Comment (tn3rb) [inline core/domain/services/admin/events/editor/EventEntityRelations.php:139]: looks like you really ARE getting a hang of the models 🥇

File: core/domain/services/graphql/connections/TicketPricesConnection.php [hunk 10]
```diff
@@ -76,39 +76,47 @@ public static function get_connection_args($args = [])
     {
         return array_merge(
             [
-                'ticket' => [
+                'in'              => [
+                    'type'        => ['list_of' => 'ID'],
+                    'description' => esc_html__('Limit pices to the given globally unique IDs', 'event_espresso'),
```
Comment (tn3rb) [inline core/domain/services/graphql/connections/TicketPricesConnection.php:81]: ahhhh... was sooooo close to being finished without finding a single thing to comment about ☹️ 

```suggestion
                    'description' => esc_html__('Limit prices to the given globally unique IDs', 'event_espresso'),
```

File: core/domain/services/graphql/connections/TicketPricesConnection.php [hunk 11]
```diff
@@ -76,39 +76,47 @@ public static function get_connection_args($args = [])
...
+                ],
+                'idIn'            => [
+                    'type'        => ['list_of' => 'ID'],
+                    'description' => esc_html__('Limit pices to the given IDs', 'event_espresso'),
```
Comment (tn3rb) [inline core/domain/services/graphql/connections/TicketPricesConnection.php:85]:
```suggestion
                    'description' => esc_html__('Limit prices to the given IDs', 'event_espresso'),
```
